pkg/dataobj/consumer/service.go

```diff
@@ -65,59 +116,75 @@ func New(kafkaCfg kafka.Config, cfg Config, mCfg metastore.Config, bucket objsto
-		kgo.OnPartitionsAssigned(partitionLifecycler.Assign),
-		kgo.OnPartitionsRevoked(partitionLifecycler.Revoke),
+		prometheus.WrapRegistererWithPrefix("loki_dataobj_consumer_", reg),
 	)
 	if err != nil {
-		level.Error(logger).Log("msg", "failed to create consumer", "err", err)
-		return nil
+		return nil, err
 	}
-	s.consumerClient = consumerClient
-	// The consumer is what polls records from Kafka. It is responsible for
-	// fetching records for each assigned partition and passing them to their
-	// respective partition processor. We guarantee that the consumer will
-	// not fetch records for partitions before the lifecycler has registered
-	// its processor because [kgo.onPartitionsAssigned] guarantees this
-	// callback returns before the client starts polling records for any
-	// newly assigned partitions.
-	consumer := newConsumer(consumerClient.Client, logger)
-	processorLifecycler.AddListener(consumer)
-
-	s.partitionLifecycler = partitionLifecycler
-	s.consumer = consumer
-	s.Service = services.NewBasicService(nil, s.running, s.stopping)
-	return s
+	s.partitionReader = partitionReader
+
+	watcher := services.NewFailureWatcher()
+	watcher.WatchService(lifecycler)
+	watcher.WatchService(partitionInstanceLifecycler)
+	s.watcher = watcher
+
+	s.Service = services.NewBasicService(s.starting, s.running, s.stopping)
+	return s, nil
+}
+
+// starting implements the Service interface's starting method.
+func (s *Service) starting(ctx context.Context) error {
+	level.Info(s.logger).Log("msg", "starting")
+	if err := services.StartAndAwaitRunning(ctx, s.lifecycler); err != nil {
+		return fmt.Errorf("failed to start lifecycler: %w", err)
+	}
+	if err := services.StartAndAwaitRunning(ctx, s.partitionInstanceLifecycler); err != nil {
+		return fmt.Errorf("failed to start partition instance lifecycler: %w", err)
+	}
+	if err := services.StartAndAwaitRunning(ctx, s.partitionReader); err != nil {
+		return fmt.Errorf("failed to start partition reader: %w", err)
+	}
+	return nil
 }
 
 // running implements the Service interface's running method.
 func (s *Service) running(ctx context.Context) error {
-	return s.consumer.Run(ctx)
+	<-ctx.Done()
+	return nil
 }
 
 // stopping implements the Service interface's stopping method.
 func (s *Service) stopping(failureCase error) error {
 	level.Info(s.logger).Log("msg", "stopping")
-	s.partitionLifecycler.Stop(context.TODO())
-	s.consumerClient.Close()
+	ctx := context.TODO()
+	if err := services.StopAndAwaitTerminated(ctx, s.partitionReader); err != nil {
+		level.Warn(s.logger).Log("msg", "failed to stop partition reader", "err", err)
+	}
+	if err := services.StopAndAwaitTerminated(ctx, s.partitionInstanceLifecycler); err != nil {
+		level.Warn(s.logger).Log("msg", "failed to stop partition instance lifecycler", "err", err)
+	}
+	if err := services.StopAndAwaitTerminated(ctx, s.lifecycler); err != nil {
+		level.Warn(s.logger).Log("msg", "failed to stop lifecycler", "err", err)
```

Comment:
this is the part where I wonder a bit how we make sure we don't leave unconsumed records in a partition.

should we stop the lifecyclers first? I think that results in them removing themselves from the ring, and then do we need to wait an amount of time before stopping the partition reader?

Reply:
We will need to implement the PrepareDownscale handlers like the ingesters: https://github.com/grafana/loki/blob/3a0480698c05e9491fc83b4efc86b65d43a21100/pkg/ingester/downscale.go#L14

This will tell the dataobj consumers to become inactive in the ring, which will stop distributors pushing records to their partitions. We can then use Delayed Downscales to allow the partition enough time to become idle and idle flush before shutting down (https://github.com/grafana/rollout-operator?tab=readme-ov-file#delayed-scaledown).